src/main/java/com/google/sps/servlets/queue/RemoveFromQueue.java

```diff
@@ -0,0 +1,110 @@
+package com.google.sps.servlets.queue;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.FirebaseAuthException;
+import com.google.firebase.auth.FirebaseToken;
+import com.google.sps.firebase.FirebaseAppManager;
+import com.google.sps.workspace.WorkspaceFactory;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.ConcurrentModificationException;
+import java.util.concurrent.ExecutionException;
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/remove-from-queue")
+public class RemoveFromQueue extends HttpServlet {
+  private FirebaseAuth authInstance;
+  private DatastoreService datastore;
+  private WorkspaceFactory factory;
+
+  @Override
+  public void init(ServletConfig config) throws ServletException {
+    try {
+      authInstance = FirebaseAuth.getInstance(FirebaseAppManager.getApp());
+      factory = WorkspaceFactory.getInstance();
+    } catch (IOException e) {
+      throw new ServletException(e);
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    // navigate to /_ah/admin to view Datastore
+
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    try {
+      String classCode = request.getParameter("classCode").trim();
+
+      // Get student info
+      String idToken = request.getParameter("idToken");
+      FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
+      String studentID = decodedToken.getUid();
+
+      int retries = 10;
+      while (true) {
+        Transaction txn = datastore.beginTransaction();
+        try {
+          // Retrive class entity
+          Key classKey = KeyFactory.stringToKey(classCode);
+          Entity classEntity = datastore.get(txn, classKey);
+
+          // Get queue
+          ArrayList<EmbeddedEntity> queue =
+              (ArrayList<EmbeddedEntity>) classEntity.getProperty("studentQueue");
+          EmbeddedEntity delEntity =
+              queue.stream().filter(elem -> elem.hasProperty(studentID)).findFirst().orElse(null);
+
+          // Delete workspace
+          EmbeddedEntity studentInfo = (EmbeddedEntity) delEntity.getProperty(studentID);
```

Comment:
What if the student is being helped? Should they be removed from that list?

Reply:
Students can only take themselves off the queue if they aren't already being helped (still waiting)